### PR TITLE
init: serde feature

### PR DESCRIPTION
**Pull request by b1ek**
ref amber-lang/amber#509

### CODE REVIEW

**Comment by Ph0enixKM** [comment]
I see why this might be useful. Could you @b1ek highlight the advantages that inspired you to create this PR? I'm just curious what exact use case you want to fulfill

**Comment by b1ek** [comment]
to serialize `Message`s for the amber integrated mode. potentially useful for a very wide range of other things also

**Comment by Mte90** [comment]
@b1ek there are conflicts in this PR

**Review by Ph0enixKM — APPROVED** [review]

**Comment by Ph0enixKM** [comment]
I want others to take a look as well

File: src/compiling/lexing/lexer.rs [hunk 1]
```diff
@@ -28,6 +32,7 @@ pub struct Lexer {
     pub scoping_mode: ScopingMode,
 }
 
+#[cfg_attr(feature = "serde", derive(Serialize))]
```
Comment (hdwalters) [inline src/compiling/lexing/lexer.rs:35]: Should this be `derive(Serialize, Deserialize)` like all the other `cfg_attr` annotations?

File: src/compiling/lexing/reader.rs [hunk 2]
```diff
@@ -1,11 +1,15 @@
...
+
 const BEGINNING: (usize, usize) = (0, 1);
 
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
 pub enum ReadMode {
     History,
     Future
 }
 
+#[cfg_attr(feature = "serde", derive(Serialize))]
```
Comment (hdwalters) [inline src/compiling/lexing/reader.rs:12]: Should this be `derive(Serialize, Deserialize)` like all the other `cfg_attr` annotations?

**Review by hdwalters — CHANGES_REQUESTED** [review]

**Comment by hdwalters** [comment]
I'm not that familiar with feature support in Rust crates, but it looks like you're creating a new feature in the `heraclitus` crate, which in turn enables serialization/deserialization feature in the `serde` crate. Can I assume you've confirmed that the `amber` project is successfully built against your local `heraclitus` changes, both with and without this feature enabled?

**Comment by b1ek** [comment]
> Can I assume you've confirmed that the `amber` project is successfully built against your local `heraclitus` changes, both with and without this feature enabled?

that is correct, i tested it before posting the PR

i just tested it now, it breaks some dependency versions with it enabled, but nothing that can't be fixed in 5 minutes

**Review by hdwalters — APPROVED** [review]

**Pull request merged**